### PR TITLE
[SPARK-26196][SPARK-26281][WebUI] Total tasks title in the stage page is incorrect when there are failed or killed tasks and update duration metrics

### PR DESCRIPTION
**Pull request by shahidki31**
## What changes were proposed in this pull request?
This PR fixes 3 issues
1) Total tasks message in the tasks table is incorrect, when there are failed or killed tasks
2) Sorting of the "Duration" column is not correct
3) Duration in the aggregated tasks summary table and the tasks table and not matching. 

Total tasks  = numCompleteTasks +  numActiveTasks + numKilledTasks + numFailedTasks;

Corrected the duration metrics in the tasks table as executorRunTime based on the PR https://github.com/apache/spark/pull/23081

## How was this patch tested?
test step:
1)
```
bin/spark-shell 
scala > sc.parallelize(1 to 100, 10).map{ x => throw new RuntimeException("Bad executor")}.collect()
```
screenshot from 2018-11-28 07-26-00

After patch:
screenshot from 2018-11-28 07-24-31

2)  Duration metrics:
Before patch:
screenshot from 2018-12-06 03-25-14

After patch:
screenshot from 2018-12-06 03-23-14

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @gengliangwang

**Review by gengliangwang — APPROVED** [review]
LGTM. Please add tag [WebUI] in the PR title

**Comment by gengliangwang** [comment]
cc @tgravescs @pgandhi999

**Comment by tgravescs** [comment]
reviewing, to clarify @pgandhi999 question, I assume the screenshots above are cut off and the task table is really showing 7 total failed tasks?

**Comment by shahidki31** [comment]
@tgravescs Yes. Task table is showing all the tasks. Sorry, I didn't put the whole screen shot.

**Review by pgandhi999 — APPROVED** [review]

File: core/src/main/resources/org/apache/spark/ui/static/stagepage.js [hunk 1]
```diff
@@ -610,7 +610,8 @@ $(document).ready(function () {
                 $("#accumulator-table").DataTable(accumulatorConf);
 
                 // building tasks table that uses server side functionality
-                var totalTasksToShow = responseBody.numCompleteTasks + responseBody.numActiveTasks;
+                var totalTasksToShow = responseBody.numCompleteTasks + responseBody.numActiveTasks +
```
[the remark this answers is not in the document]
Comment (shahidki31) [reply core/src/main/resources/org/apache/spark/ui/static/stagepage.js]: Yes @pgandhi999 . All the failed tasks are displaying in the tasks table. Thanks for your great work.

**Review by tgravescs — APPROVED** [review]
+1, thanks @shahidki31 !

**Comment by shahidki31** [comment]
Hi @pgandhi999 , There is a small nit there. Recently we fixed a bug related to the "duration" metrics in the tasks table (see https://github.com/apache/spark/pull/23081), but that hasn't reflected here. I have updated the code

File: core/src/main/resources/org/apache/spark/ui/static/stagepage.js [hunk 2]
```diff
@@ -661,8 +662,8 @@ $(document).ready(function () {
                         {data : "launchTime", name: "Launch Time", render: formatDate},
                         {
                             data : function (row, type) {
-                                if (row.duration) {
-                                    return type === 'display' ? formatDuration(row.duration) : row.duration;
+                                if (row.taskMetrics && row.taskMetrics.executorRunTime) {
+                                    return type === 'display' ? formatDuration(row.taskMetrics.executorRunTime) : row.taskMetrics.executorRunTime;
```
Comment (pgandhi999) [inline core/src/main/resources/org/apache/spark/ui/static/stagepage.js]: Ok, that makes sense. However, you also need to update the search filter method to not search on duration in StagesResource.scala here:
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/api/v1/StagesResource.scala#L213.

**Review by pgandhi999 — COMMENTED** [review]
My guess is that this might break the sorting on the Duration column in the tasks table as we are using the value of the name field in the column data to match with the appropriate name of the data field in the server data below in StagesResource:

`withUI(_.store.taskList(stageId, stageAttemptId, pageStartIndex, pageLength,
      indexName(columnNameToSort), isAscendingStr.equalsIgnoreCase("asc")))
  }`
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/api/v1/StagesResource.scala#L178. 
So as far as I can think, I expect that if you sort the Duration column, it will perform sorting on row.duration instead of row.taskMetrics.executorRunTime, thus, not getting the desired results.

Can you please verify whether sorting on Duration column works as expected and fix if necessary? Thank you.

File: core/src/main/resources/org/apache/spark/ui/static/stagepage.js (hunk 2 above)
Comment (shahidki31) [reply core/src/main/resources/org/apache/spark/ui/static/stagepage.js]: Thanks removed.

**Comment by shahidki31** [comment]
> So as far as I can think, I expect that if you sort the Duration column, it will perform sorting on row.duration instead of row.taskMetrics.executorRunTime, thus, not getting the desired results.

@pgandhi999 Actually we have mapped the "Duration" to "executorRunTime" for sorting in the PR which I have mentioned above. So, after this PR the "Duration" is sorting correctly

**Comment by pgandhi999** [comment]
LGTM

**Comment by shahidki31** [comment]
cc @tgravescs @srowen

**Review by srowen — APPROVED** [review]

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master. @shahidki31 does this need to go in branch 2.4, 2.3? I tried back porting it, but looks like a lot of the affected code didn't exist in 2.4. If the fix can or should also be back-ported and you're willing, you're welcome to open another PR against 2.4.

**Comment by shahidki31** [comment]
Thanks @srowen. This issue happens only in master branch.

Thank you all for the review and comments

**Pull request closed without merging**